### PR TITLE
Replace models with Workflow models

### PR DESCRIPTION
**Pull request by AlmightyYakob**
This PR accomplishes the following main things:
* Replace the existing models with workflow specific models
* Create an views for interfacing with these workflows from the API
* Layout the ground work for running these workflows

### CODE REVIEW

**Review by banesullivan — APPROVED** [review]
This is looking really, really awesome! Excellent work here. @AlmightyYakob!

The tests are looking great and I appreciate that you used ViewSets for the API

Only minor changes suggested and one comment on step insertion that can be addressed later.

Good to merge!

**Comment by banesullivan** [comment]
Also, @AlmightyYakob would you go ahead and move this algorithms app to the top level of the repo per https://docs.djangoproject.com/en/3.2/intro/reusable-apps/#packaging-your-app

I also think we should rename it to `django-rgd-workflow` with `rgd_workflow` as the module name

File: rdoasis/algorithms/models/workflow.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    # Give related name of child_links, because for any entry with parent=node,
+    # all links from node.child_links have the form (parent=node, child=child_node)
+    parent = models.ForeignKey('WorkflowStep', related_name='child_links', on_delete=models.CASCADE)
+
+    # Give related name of parent_links, because for any entry with child=node,
+    # all links from node.parent_links have the form (parent=parent_node, child=node)
+    child = models.ForeignKey('WorkflowStep', related_name='parent_links', on_delete=models.CASCADE)
+
+    # The distance between the parent and child (direct child = 1, grandchild = 2, etc.)
+    distance = models.PositiveIntegerField()
```
[the remark this answers is not in the document]
Comment (AlmightyYakob) [reply rdoasis/algorithms/models/workflow.py:77]: I don't think computing `distance` dynamically is possible/feasible. The benefit that's gained from having the `distance` field is precisely that it's calculated ahead of time, allowing us to leverage that field for many different queries.

You're correct that insertion between steps isn't currently handled. However, I don't think adding such a function would be too complicated. I'll have to think about the specifics of this, but I think it's important to have this nailed down, so I'll try to add it into this PR. If it somehow spirals out of control, I might put it off the the time being.

**Comment by AlmightyYakob** [comment]
> With some after thought, @AlmightyYakob would you go ahead and move this algorithms app to the top level of the repo per [docs.djangoproject.com/en/3.2/intro/reusable-apps/#packaging-your-app](https://docs.djangoproject.com/en/3.2/intro/reusable-apps/#packaging-your-app)

Sure thing

> I also think we should rename it to `django-rgd-workflow` with `rgd_workflow` as the module name

I'm fine with that, but if we plan on upstreaming this into https://github.com/ResonantGeoData/ResonantGeoData, would the rename cause any issues there? I suppose not, as we'd just be changing where the code itself lived, but just want to be sure.

File: rdoasis/algorithms/models/workflow.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    links.update(distance=models.F('distance') - 1)
+
+
+class Workflow(TimeStampedModel):
+    """A model for organizing the running of workflow steps."""
+
+    name = models.CharField(max_length=255, unique=True)
+    collection = models.OneToOneField(
+        Collection, on_delete=models.PROTECT, default=create_default_workflow_collection
+    )
```
[the remark this answers is not in the document]
Comment (AlmightyYakob) [reply rdoasis/algorithms/models/workflow.py:193]: Yeah, after our discussions around this, I agree this should be swapped out for something else. However, I suppose it's fine for now.

**Comment by AlmightyYakob** [comment]
> With some after thought, @AlmightyYakob would you go ahead and move this algorithms app to the top level of the repo per [docs.djangoproject.com/en/3.2/intro/reusable-apps/#packaging-your-app](https://docs.djangoproject.com/en/3.2/intro/reusable-apps/#packaging-your-app)
> 
> I also think we should rename it to `django-rgd-workflow` with `rgd_workflow` as the module name

Also, is the `core` app still needed? And if so, how should that and the `algorithms` app be coordinated?

File: rdoasis/algorithms/models/workflow.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (banesullivan) [reply rdoasis/algorithms/models/workflow.py:77]: Ah, okay. I think I see what `distance` needs to be stored field.

> ... I might put it off the the time being.

Definitely don't hesitate to just put that off for now. I'd like to get the core features merged in, then handle these other use cases as smaller PRs

**Comment by AlmightyYakob** [comment]
> I don't think there would be any issues from the rename. It would be consistent in the fact that this is another Django app built on top of `django-rgd` core

:+1: 

> Yes, the `core` app is still needed for project-specific migrations and management commands and should remain in the Django Project directory. What do you mean by coordination? the algorithms/workflow app is completely independent of `core` and at this time, `core` doesn't have anything linking to `algorithms`

I essentially wanted to make sure of this. I'll go ahead with the reorg and rename.

File: rdoasis/algorithms/models/workflow.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlmightyYakob) [reply rdoasis/algorithms/models/workflow.py:77]: After some initial investigation, I'm going to put that off for the time being.

File: rdoasis/algorithms/urls.py [hunk 3]
```diff
@@ -1,54 +1,23 @@
...
-                'queryset': model.objects.all(),
-                'serializer_class': ser,
-                'filter_backends': [DjangoFilterBackend],
-                'filterset_fields': utility.get_filter_fields(model),
-            },
-        )
-        viewset_class.__doc__ = model.__doc__
-        router.register('api/%s' % (model_name.lower()), viewset_class)
+router = ExtendedSimpleRouter()
+workflow_routes = router.register('workflow', WorkflowViewSet)
```
[the remark this answers is not in the document]
Comment (AlmightyYakob) [reply rdoasis/algorithms/urls.py]: cd769b7

**Pull request merged**